hack/verify-flags-underscore.sh make sure we always use - in flags not _

eparis opened the pull request:
This works by defining two 'static' lists in hack. The first is the list
of all flags in the project which use a `-` or an `_` in their name. All
files being processed by verify-flags-underscore.py (or all files in the
repo if no filename arguments are given) will be searched for flag
declaration using a simple regex. Its not super smart. If a flag is
found which is not in the static list it will complain/reject the commit
until a human adds it to the list. If we do not keep a static list of
flags it takes >.2 seconds to find them 'all' at runtime. Since this is
run in pre-commit saving every part of a second helps.

After it finds all of the flags it runs all of the arguments (or all
files in repo if no arguments) looking for usage of those flags which
includes an `_`. There are lots of places where these are false
positives. For example we have a flag named oom-adj-score but the kernel
calls it oom_adj_score. To handle this we keep a second 'whitelist' of
lines which are allowed to use these flag names with an `_`.

Running the entire git repo looking for flags in every golang file and
looking in every single file for bad usage takes about 8.75 seconds.

Running it in the precommit hook where we only check things that changed
takes about .06 seconds.

eparis committed:
Verify all flag usage does not use _

This works by defining two 'static' lists in hack. The first is the list
of all flags in the project which use a `-` or an `_` in their name. All
files being processed by verify-flags-underscore.py (or all files in the
repo if no filename arguments are given) will be searched for flag
declaration using a simple regex. Its not super smart. If a flag is
found which is not in the static list it will complain/reject the commit
until a human adds it to the list. If we do not keep a static list of
flags it takes >.2 seconds to find them 'all' at runtime. Since this is
run in pre-commit saving every part of a second helps.

After it finds all of the flags it runs all of the arguments (or all
files in repo if no arguments) looking for usage of those flags which
includes an `_`. There are lots of places where these are false
positives. For example we have a flag named oom-adj-score but the kernel
calls it oom_adj_score. To handle this we keep a second 'whitelist' of
lines which are allowed to use these flag names with an `_`.

Running the entire git repo looking for flags in every golang file and
looking in every single file for bad usage takes about 8.75 seconds.

Running it in the precommit hook where we only check things that changed
takes about .06 seconds.

eparis commented:
@bgrant0607 I don't love how it turned out, but not sure how to make it a whole lot better...

bgrant0607 commented:
Thank you! The whitelist was a good idea.

bgrant0607 commented:
cc @krousey @mikedanese in case @zmerlynn doesn't have time to review right now

thockin commented:
How long does this new presubmit take to run?    Are you spending the gains you made (but haven't committed yet :)

thockin commented:
Add to shippable and travis, please

eparis commented:
thockin, yes.  It costs about 1s in pre-commit.   (10s to run on the whole repo)

zmerlynn commented:
Shuffling this one off, sorry, was OOO on Friday and catching up.

krousey commented on hack/verify-flags-underscore-exceptions.txt:
```diff
@@ -0,0 +1,453 @@
+contrib/for-tests/mount-tester/mt.go:	flag.StringVar(&newFilePath0666, "new_file_0666", "", "Path to write to and read from with perm 0666")
+contrib/for-tests/mount-tester/mt.go:	flag.StringVar(&newFilePath0777, "new_file_0777", "", "Path to write to and read from with perm 0777")
+contrib/mesos/pkg/controllermanager/controllermanager.go:	fs.BoolVar(&s.UseHostPortEndpoints, "host_port_endpoints", s.UseHostPortEndpoints, "Map service endpoints to hostIP:hostPort instead of podIP:containerPort. Default true.")
+contrib/mesos/docs/ha.md:- `--km_path` or else (`--executor_path` and `--proxy_path`) should reference non-local-file URI's and must be identicial across schedulers.
```
Why do there need to be exceptions for files that aren't Go? It looks as though you filter out non-Go files below.

eparis replied:
We filter out non-go to try to find all flag names.  We run ALL files looking for bad usage of such flags.

krousey replied:
Ah I see. So the point of this is to ensure proper usage of flags. I thought it was to ensure we didn't create bad flags to begin with.

eparis replied:
It does both, as the second sweep looking for 'usage' also picks up the 'declaration'

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,177 @@
+    a = ['Godeps', 'third_party']
+    for f in files:
+        if any(x in f for x in a):
+            print("shit %s" % f)
```
don't forget to delete this line

eparis replied:
woah!!! fixed.

krousey commented:
@eparis Any plans to expand it to catch the non `*Var(...)` flags? It might be less error prone to write this in Go and use Go's parser and AST to find these things instead of using regular expressions. go vet has a similar feature where it finds calls to `Printf` (and others) and checks that the formatting string and other arguments match up. It could be a bit of effort though...

eparis commented:
Updated the initial comment to better describe the work.

- pre-commit hook now takes < .1 seconds in most cases
- added to travis and shippable (still won't work as this is WIP)
- I picked up VarP versions
- looking at using the gofmt AST might get me something, but we don't add flags like mad, so the whitelist of flags should get us a long way...

- This is still a WIP so I know travis/shippable are going to fail (on this check, so we can see how it works!)

PTAL as I think the code is now ready, even if I need to/should make other fixes in the tree before commiting.

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+            continue
+        f = open(pathname, 'r')
+        for line in f.read().splitlines():
+            match = fre.search(line)
```
Are we at all worried about a multiline flag definition?

```go
func init() {
	flag.BoolVar(&reallyLongAndComplicatedVariableName,
		"a_bad_flag", false, "usage")
}
```

eparis replied:
I wasn't, but I think I can make it work...

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+        CHUNKSIZE = 1024
+        while 1:
+            chunk = f.read(CHUNKSIZE)
+            if '\0' in chunk: # found null byte
```
So I take it we don't care about UTF-16? :smile: (don't worry about it)

eparis commented:
fix multi-line flag declarations (turns out there were 4-5)

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+            continue
+        filename = out[0]
+        line = out[1]
+        exceptions[(filename, line)] = True
```
You could use a `set` instead of `dict` here.

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+    pathname = os.path.join(rootdir, "hack/verify-flags/known-flags.txt")
+    f = open(pathname, 'r')
+    for line in f.read().splitlines():
+        flags[line] = True
```
You could use a `set` instead if you need a unique container without caring about having it map to something.

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+    return flags
+
+def flags_to_re(flags):
+    # turn the list of all flags we found into a regex string which will find both - and _ version
```
This seems more like a doc string than a comment. Could you make it one?

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,210 @@
+        return False
+    if "_" in m.group(0):
+        return True
+    return False
```
```python
return m and "_" in m.group(0)
```
?

eparis replied:
Would be the same thing, but I like my middle school coding style

krousey commented on hack/verify-flags-underscore.py:
```diff
@@ -0,0 +1,213 @@
+    for pathname in files:
+        relname = os.path.relpath(pathname, rootdir)
+        f = open(pathname, 'r')
+        for line in f.read().splitlines():
```
Just to be certain, this is fine to operate on individual lines because we're matching explicit flag names and they shouldn't split across lines anyways?

eparis replied:
correct, i can probably change it to use findall as well.

eparis replied:
nah, since the 'exceptions' are per line and flag names can't span lines, I'm not going to change this bit and it should be fine.

krousey commented:
@eparis It looks good with some minor nits that you can ignore if you want. Let me know when you thing this is no longer a WIP and should be ready to go in.

eparis commented:
ok.  /me gets to work trying to clean up the remaining `-` v `_` usages in the tree.

eparis commented:
Needs rebase after
- #12549
- #12550
- #12551

eparis commented:
Hopefully this commit is now ready for final review and merge.  I got a whole bunch of `_` removed from the repo!

krousey commented on hooks/pre-commit:
```diff
@@ -53,6 +53,22 @@ else
 fi
 echo "${reset}"
 
+allfiles=($(git diff --cached --name-only --diff-filter ACM | grep -v -e "third_party" -e "Godeps"))
```
I see this pattern a lot. Would it make sense to get the list from git once and then apply subfilters? Probably not this PR.

bgrant0607 commented:
Apparently is broken at head:
```
  Found illegal 'flag' usage. If this is a false positive add the following line(s) to hack/verify-flags/exceptions.txt:

cluster/libvirt-coreos/user_data.yml:    advertise-client-urls: http://${public_ip}:2379
cluster/libvirt-coreos/user_data.yml:    initial-advertise-peer-urls: http://${public_ip}:2380
cluster/libvirt-coreos/user_data.yml:    listen-peer-urls: http://${public_ip}:2380
```

eparis commented:
will send a PR.

eparis commented:
https://github.com/kubernetes/kubernetes/pull/12687
